src/v/cloud_storage_clients/abs_client.cc

```diff
@@ -0,0 +1,600 @@
+  ss::lw_shared_ptr<const cloud_roles::apply_credentials> apply_credentials)
+  : _requestor(conf, std::move(apply_credentials))
+  , _client(conf, &as, conf._probe, conf.max_idle_time)
+  , _probe(conf._probe) {}
+
+ss::future<> abs_client::stop() { return _client.stop(); }
+
+void abs_client::shutdown() { _client.shutdown(); }
+
+template<typename T>
+ss::future<result<T, error_outcome>> abs_client::send_request(
+  ss::future<T> request_future,
+  const bucket_name& bucket,
+  const object_key& key) {
+    using namespace boost::beast::http;
+
+    auto outcome = error_outcome::retry;
+
+    try {
+        co_return co_await std::move(request_future);
+    } catch (const abs_rest_error_response& err) {
+        _probe->register_failure(err.code());
+        if (err.code() == abs_error_code::_unknown) {
+            vlog(
+              abs_log.error,
+              "Accessing storage account {}, unexpected REST API error "
+              "detected: code={}, http_code={}; {}",
+              bucket,
+              err.code_string(),
+              err.http_code(),
+              err.message());
+            outcome = error_outcome::fail;
+        } else if (
+          err.http_code() == status::forbidden
+          || err.http_code() == status::unauthorized) {
+            vlog(
+              abs_log.error,
+              "Received {} error response from ABS. This indicates "
+              "misconfiguration of ABS and/or Redpanda: {}",
+              err.code_string(),
+              err.message());
+            outcome = error_outcome::fail;
+        } else if (
+          err.code() == abs_error_code::container_being_disabled
+          || err.code() == abs_error_code::container_being_deleted
+          || err.code() == abs_error_code::container_not_found) {
+            vlog(
+              abs_log.error,
+              "Received {} error response from ABS. This indicates "
+              "that your container is not available. Remediate the issue for "
+              "uploads to resume: {}",
+              err.code_string(),
+              err.message());
+            outcome = error_outcome::fail;
+        } else if (err.code() == abs_error_code::blob_not_found) {
+            // Unexpected 404s are logged by 'request_future' at warn
+            // level, so only log at debug level here.
+            vlog(abs_log.debug, "BlobNotFound response received {}", key);
+            outcome = error_outcome::key_not_found;
+        } else if (err.code() == abs_error_code::server_busy) {
+            vlog(
+              abs_log.warn,
+              "{} response received for container {}",
+              err.code_string(),
+              bucket);
+            outcome = error_outcome::retry_slowdown;
+        } else {
+            vlog(
+              abs_log.trace,
+              "Received {} retryable error resopnse from ABS: {}",
+              err.code_string(),
+              err.message());
+            outcome = error_outcome::retry;
```

Comment:
I'm suggesting to change this to `error_outcome::fail` because it's a wildcard and some non-retryable REST API error could easily trigger this. It's better to mark retry cases explicitly. Probably, apart from `server_busy` nothing should be retryable.

Reply:
I'm not sure. The error handling logic is a bit different from the S3 client.

```
enum class abs_error_code {
    _unknown,
    blob_not_found,
    authentication_failed,
    server_busy,
    internal_error,
    system_in_use,
    account_being_created,
    resource_already_exists,
    blob_already_exists,
    invalid_blob,
    pending_copy_operation,
    snapshot_present,
    blob_being_rehydrated,
    container_being_disabled,
    container_being_deleted,
    container_not_found
};
```

This is how the error code type is defined. Note that only a sub-set of the possible error codes were explicitly defined and everything else will be parsed as `_unknown` (maps to `error_outcome::fail`. `container_being_disabled`, `container_being_deleted`, `container_not_found` and `authentication_failed` map to `error_outcome::fail`, `blob_not_found` maps to `error_outcome::key_not_found`, `server_busy` maps to `error_outcome::retry_slowdown`. The rest are potentially transient errors. For example, if the account is being created, I think it makes sense to retry.

I'm open to changing mappings for specific error codes if you think it makes sense though.

Reply:
Yes, I saw the type. It makes me thing what will we see in the logs in case of `_unknown` error. 

Regarding the retry behavior. If you will return 'fail' when account is being created it will be retried anyway. The difference between `retry_slowdown` and `fail` is subtle but important. The `retry_slowdown` increments corresponding metric that shows amount of throttling in the cloud. Also, `retry_slowdown` is retried using exponential backoff and `fail` is not. After receiving `fail` the `remote` will log and return an error and then `ntp_archiver` will retry. These two cases are different from observability standpoint. Also, it doesn't make sense to use exponential backoff to retry after `account_being_created` and some other errors.

One example would be the segment download in `remote_segment`. If you're reading from tiered-storage and some unexpected error occurs (which isn't `server_busy`) you want the read path to return an error instead of blocking and waiting for 30 seconds (which is a default segment download timeout).

Reply:
Makes sense. I'll think about it some more. My thinking now is that we should have `server_busy` and `internal_error` as `retry_slowdown` and the rest should probably be `fail`.

Reply:
`internal_error` should probably be a `fail` since the server doesn't tells you explicitly that you have to wait for a while. I also think about that difference in throttling behavior between ABS and S3. IIRC, ABS doesn't throttle the app the same way S3 does. S3 replies with `SlowDown` response and expects you to use exponential backoff and retry.  Does `server_busy` behaves the same way in ABS? And what happens when you're running out of traffic in your storage account? Does ABS expects you to use exponential backoff and retry or it will always give you an error until the next billing period will start?

Reply:
> S3 replies with SlowDown response and expects you to use exponential backoff and retry. Does server_busy behaves the same way in ABS?

Azure expects you to apply exponential backoff when receiving a `server_busy` (HTTP code 503) or a `timeout_error` (HTTP code 500) [[source](https://learn.microsoft.com/en-us/azure/storage/blobs/storage-performance-checklist#timeout-and-server-busy-errors)].

> And what happens when you're running out of traffic in your storage account?

You'll get `server_busy` responses [[source](https://learn.microsoft.com/en-us/rest/api/storageservices/common-rest-api-error-codes); search for ServerBusy].

> Does ABS expects you to use exponential backoff and retry or it will always give you an error until the next billing period will start?

Clients are expected to back-off exponentially.

I also had a look at how the clients provided by Microsoft handle retries. Both the C# and Python client retry on the following HTTP codes: `[408, 429, 500, 502, 503, 504]`. This is more error codes than I expected and a number of these are not even mentioned in their docs as possible. My guess is that they have older server versions serving user requests and this list caters to all of them.

I'll do `retry_slowdown` on the mentioned HTTP codes and `fail` the rest of them.

Reply:
True, we parse that XML too. The [C#](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/ResponseClassifier.cs#L20) and [python](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/core/azure-core/azure/core/pipeline/policies/_retry.py#L71-L72) clients decide whether to retry by looking only at the HTTP code though.

Going through the [docs](https://learn.microsoft.com/en-us/rest/api/storageservices/common-rest-api-error-codes), those HTTP codes map to the following ABS error codes: `InternalError` (500), `OperationTimedOut` (500), `ServerBusy` (503). The recommendation is to exponentially retry all of these, so we should be good. The remaining HTTP codes, `[408, 429, 502, 503, 504]`, don't map to any ABS error code according to the docs. Like I said, some older server versions probably returned them (e.g. HTTP code 429 instead of `ServerBusy` with HTTP code 503). I don't think it's harmfull to keep them.

Reply:
Updated the error handling as described in the previous comment. See https://github.com/redpanda-data/redpanda/compare/941c2cc83645b4fbcf6724407cdd3b477e6f3f1f..4d2353e885b4100a1fe7ce330c14d987238ad8d9.